### PR TITLE
Add recipe for xgboost

### PR DESCRIPTION
**Pull request by nvkevlu**
Add recipe for xgboost.

### Description

Add recipe for xgboost.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/xgboost/fedxgb/README.md [hunk 1]
```diff
@@ -3,8 +3,43 @@ Several mechanisms have been proposed for training an XGBoost model in a federat
 In these examples, we illustrate the use of NVFlare to carry out *horizontal* federated learning using two approaches: histogram-based collaboration and tree-based collaboration.
 And *vertical* federated learning using histogram-based collaboration.
 
+## What's New with Recipe API
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/README.md]: This is not the place to explain Recipe API.  I am not sure we need what's new part.

File: examples/advanced/xgboost/fedxgb/README.md [hunk 2]
```diff
@@ -3,8 +3,43 @@ Several mechanisms have been proposed for training an XGBoost model in a federat
...
+
+After (Recipe API):
+```python
+from nvflare.app_opt.xgboost.recipes import XGBHistogramRecipe
+
+recipe = XGBHistogramRecipe(
+    name="xgb_higgs",
+    min_clients=2,
+    num_rounds=100,
+    algorithm="histogram_v2",
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/README.md]: everything else has train_scripts or something, why this one called algorithm ?

File: examples/advanced/xgboost/fedxgb/README.md [hunk 3]
```diff
@@ -128,23 +163,82 @@ and can even make the accuracy worse if subsample rate is too low (e.g. 0.05).
...
 ### Horizontal Experiments
 The following cases will be covered:
-- Histogram-based collaboration based on uniform data split for 2 / 5 clients
+- Histogram-based collaboration based on uniform data split for 2 / 5 clients (using **Recipe API**)
 - Tree-based collaboration with cyclic training based on uniform / exponential / square data split for 5 / 20 clients
 - Tree-based collaboration with bagging training based on uniform / exponential / square data split for 5 / 20 clients w/ and w/o scaled learning rate
 
-Histogram-based experiments can be run with:
-```
+#### Histogram-Based (Recipe API)
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/README.md]: Remove Recipe API part. As everything will be recipe API based

File: examples/advanced/xgboost/fedxgb/higgs_data_loader.py [hunk 4]
```diff
@@ -72,6 +72,6 @@ def load_data(self):
         x_valid, y_valid, total_valid_data_num = _read_higgs_with_pandas(
             data_path=data_path, start=valid_index["start"], end=valid_index["end"]
         )
-        dmat_valid = xgb.DMatrix(x_valid, label=y_valid, data_split_mode=self.data_split_mode)
+        dmat_valid = xgb.DMatrix(x_valid, label=y_valid)
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/higgs_data_loader.py]: we should keep data_split_mode

File: examples/advanced/xgboost/fedxgb/job.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        default="/tmp/nvflare/dataset/xgboost_higgs",
+        help="Path to dataset files for each site",
+    )
+    parser.add_argument("--site_num", type=int, default=2, help="Total number of sites")
+    parser.add_argument("--round_num", type=int, default=100, help="Total number of training rounds")
+    parser.add_argument(
+        "--training_algo",
+        type=str,
+        default="histogram_v2",
+        choices=["histogram", "histogram_v2"],
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/job.py]: Not related to this PR but do we need the "histogram" mode anymore? @ZiyueXu77 @chesterxgchen or should we remove it? also the code along with it

File: examples/advanced/xgboost/fedxgb/job_vertical.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    job = FedJob(name=job_name, min_clients=args.site_num)
+
+    # PSI Controller
+    controller = DhPSIController()
+    job.to_server(controller)
+
+    # PSI Executor
+    executor = PSIExecutor(psi_algo_id="dh_psi")
+    job.to_clients(executor, id="psi_executor", tasks=["PSI"])
+
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/job_vertical.py]: we have a recipe for DhPSI, please refer to "examples/advanced/psi/user_email_match/" and use it

File: tests/integration_test/test_xgb_secure_recipe.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline tests/integration_test/test_xgb_secure_recipe.py]: what is secure recipe?

File: examples/advanced/xgboost/fedxgb_secure/job.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            secure=args.secure,
+            xgb_params=xgb_params,
+        )
+
+        # Add data loaders to each client
+        # Note: CSVDataLoader automatically handles client-specific data loading.
+        # Even though we pass the same folder path, each client will load its own data
+        # from {dataset_path}/{client_id}/ at runtime (e.g., site-1 loads from site-1/ subdirectory)
+        for i in range(1, args.site_num + 1):
+            dataloader = CSVDataLoader(folder=dataset_path)
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb_secure/job.py]: since they are all the same, should we just pass this in the constructor?

File: examples/advanced/xgboost/fedxgb_secure/job.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        "objective": "binary:logistic",
+        "eval_metric": "auc",
+        "tree_method": args.tree_method,
+        "nthread": args.nthread,
+    }
+
+    # Create recipe based on data split mode
+    if args.data_split_mode == "horizontal":
+        # Horizontal histogram-based XGBoost
+        recipe = XGBHistogramRecipe(
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb_secure/job.py]: Can these 2 recipes be merged into 1?
they only differ by split mode?

File: nvflare/app_opt/xgboost/recipes/vertical.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+                "eta": 0.1,
+                "objective": "binary:logistic",
+                "eval_metric": "auc",
+                "tree_method": "hist",
+                "nthread": 16,
+            }
+
+        # Auto-generate client ranks if not provided and secure=True
+        if secure and client_ranks is None:
+            client_ranks = {f"site-{i}": i - 1 for i in range(1, min_clients + 1)}
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/recipes/vertical.py]: I think we should not do this inside a recipe `__init__`

File: nvflare/app_opt/xgboost/recipes/vertical.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+
+        Note:
+            The dataloader must be configured with:
+            - data_split_path: Path to this client's feature data
+            - psi_path: Path to PSI intersection file for this client
+            - id_col: Column name for sample IDs
+            - label_owner: Same as recipe's label_owner
+            - train_proportion: Train/validation split ratio
+        """
+        from nvflare.app_opt.xgboost.histogram_based_v2.fed_executor import FedXGBHistogramExecutor
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/recipes/vertical.py:239]: let's in another PR add some deprecation warning in the V1 and in 2.8.0 we remove V1 and rename V2.

File: examples/advanced/xgboost/fedxgb/xgb_fl_job_horizontal.py [hunk 12]
```diff
@@ -1,260 +0,0 @@
...
-        )
-        from nvflare.app_opt.xgboost.tree_based.model_persistor import XGBModelPersistor
-
-        persistor = XGBModelPersistor(save_name="xgboost_model.json")
-        from nvflare.app_opt.xgboost.tree_based.shareable_generator import XGBModelShareableGenerator
-
-        shareable_generator = XGBModelShareableGenerator()
-        from nvflare.app_opt.xgboost.tree_based.bagging_aggregator import XGBBaggingAggregator
-
-        aggregator = XGBBaggingAggregator()
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/xgb_fl_job_horizontal.py:182]: we need the bagging recipe as well.

File: examples/advanced/xgboost/fedxgb/README.md [hunk 13]
```diff
@@ -128,23 +163,82 @@ and can even make the accuracy worse if subsample rate is too low (e.g. 0.05).
...
 
-Model accuracy curve during training can be visualized in tensorboard, 
+Model accuracy curve during training can be visualized in tensorboard,
 recorded in the simulator folder under `/tmp/nvflare/workspace/works/`.
 As expected, we can observe that all histogram-based experiments results in identical curves as centeralized training:
 ![Horizontal Histogram XGBoost Graph](./figs/histogram.png)
 
+#### Tree-Based (Bagging and Cyclic)
+
+> **_NOTE:_** Tree-based algorithms (bagging and cyclic) are not yet converted to Recipe API.
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/README.md]: why ? this should be doable?

File: examples/advanced/xgboost/fedxgb/README.md [hunk 14]
```diff
@@ -3,8 +3,43 @@ Several mechanisms have been proposed for training an XGBoost model in a federat
...
+## What's New with Recipe API
+
+The histogram-based horizontal federated learning examples now use the **Recipe API** for simplified job configuration.
+
+**Key improvements**:
+- **Simpler code**: Recipe API reduces boilerplate by ~40%
+- **Single job file**: `job.py` replaces `xgb_fl_job_horizontal.py`
+- **Cleaner structure**: Data loader moved from `src/` to root level for consistency
+- **Same functionality**: All features preserved (histogram, histogram_v2, TensorBoard tracking)
+
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/README.md]: please remove the what's new part

**Review by YuanTingHsieh — DISMISSED** [review]

File: examples/advanced/xgboost/fedxgb/README.md (hunk 13 above)
Comment (YuanTingHsieh) [reply examples/advanced/xgboost/fedxgb/README.md]: can we reuse (and/or enhance) nvflare/app_opt/xgboost/recipes/bagging.py

File: examples/advanced/xgboost/fedxgb/xgb_fl_job_horizontal.py (hunk 12 above)
Comment (YuanTingHsieh) [reply examples/advanced/xgboost/fedxgb/xgb_fl_job_horizontal.py:182]: can we reuse (and/or enhance) nvflare/app_opt/xgboost/recipes/bagging.py

File: examples/advanced/xgboost/fedxgb/README.md (hunk 2 above)
Comment (nvkevlu) [reply examples/advanced/xgboost/fedxgb/README.md]: Script-based frameworks: User provides their own training logic → need path to script
XGBoost: User selects from pre-built federated algorithms → need algorithm name

one option could be to refactor XGBoost to accept optional train_script for custom training logic, but default to built-in algorithms if not provided. This gives true parameter consistency AND future extensibility, but may require significant additional changes.

Comment (chesterxgchen) [reply examples/advanced/xgboost/fedxgb/README.md]: we only has one algorithm. @YuanTingHsieh is going to remove the old V1 version and rename the v2 to histogram. 

What are the options for the user to select ? Histogram-based vs. tree-based?  

We might need a XGBoost algorithm type ? @ZiyueXu77  what should this be called ?

File: tests/integration_test/test_xgb_secure_recipe.py (hunk 7 above)
Comment (nvkevlu) [reply tests/integration_test/test_xgb_secure_recipe.py]: secure training uses Homomorphic Encryption (HE), this converts an existing FedJob API example to use the Recipe API

File: nvflare/app_opt/xgboost/recipes/histogram.py [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+        self.job = self.configure()
+        Recipe.__init__(self, self.job)
+
+    def configure(self):
+        """Configure the federated job for XGBoost histogram-based training."""
+        # Create FedJob
+        job = FedJob(name=self.name, min_clients=self.min_clients)
+
+        # Configure controller and executor based on algorithm
+        if self.algorithm == "histogram":
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/recipes/histogram.py]: let's remove all "histogram" stuff in this example, in another PR we can remove or mark as deprecated.

File: nvflare/app_opt/xgboost/recipes/histogram.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+                xgb_params=self.xgb_params,
+                metrics_writer_id=self.metrics_writer_id,
+                use_gpus=self.use_gpus,
+            )
+            job.to_clients(executor, id="xgb_executor")
+
+        elif self.algorithm == "histogram_v2":
+            # Newer histogram-based implementation (recommended)
+            from nvflare.app_opt.xgboost.histogram_based_v2.fed_controller import XGBFedController
+            from nvflare.app_opt.xgboost.histogram_based_v2.fed_executor import FedXGBHistogramExecutor
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/recipes/histogram.py]: i see both horizontal and vertical using this, in that case lets just merge these 2 recipe, and have an arg call "axis" or "split_mode" or the name you prefer

File: examples/advanced/xgboost/fedxgb/README.md [hunk 17]
```diff
@@ -128,27 +128,146 @@ and can even make the accuracy worse if subsample rate is too low (e.g. 0.05).
...
+```bash
+python job.py \
+    --site_num 2 \
+    --round_num 100 \
+    --training_algo histogram_v2 \
+    --split_method uniform
+```
+
+**Available parameters**:
+- `--training_algo`: Choose `histogram` or `histogram_v2` (recommended)
```
Comment (YuanTingHsieh) [inline examples/advanced/xgboost/fedxgb/README.md]: lets remove all v1

File: nvflare/app_opt/xgboost/recipes/vertical.py [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+
+    @field_validator("num_rounds")
+    @classmethod
+    def check_num_rounds(cls, v):
+        if v < 1:
+            raise ValueError("num_rounds must be at least 1")
+        return v
+
+
+class XGBVerticalRecipe(Recipe):
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/xgboost/recipes/vertical.py:57]: can this be merged with XGBHistogramRecipe?

Comment (nvkevlu) [reply nvflare/app_opt/xgboost/recipes/vertical.py:57]: I thought Chester was saying:

It may sounds simple to share the code, for end users , this actually confusing. We don't mind a bit duplication

what do you think? is merging it preferred?

File: examples/advanced/xgboost/fedxgb/run_experiment_vertical.sh [hunk 19]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
-python3 xgb_fl_job_vertical_psi.py
-python3 xgb_fl_job_vertical.py
\ No newline at end of file
+# Run both PSI and training in sequence
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/run_experiment_vertical.sh]: not sure we need this sh script, just let me run one after another, no need to have a shell script

File: examples/advanced/xgboost/fedxgb/run_experiment_horizontal_histogram.sh [hunk 20]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-python3 xgb_fl_job_horizontal.py --site_num 2 --training_algo histogram --split_method uniform --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo histogram --split_method uniform --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 2 --training_algo histogram_v2 --split_method uniform --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo histogram_v2 --split_method uniform --lr_mode uniform --data_split_mode horizontal
+python3 job.py --site_num 2 --training_algo histogram --split_method uniform --data_split_mode horizontal
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/run_experiment_horizontal_histogram.sh]: these are different experiments, why do we need all these ?

File: examples/advanced/xgboost/fedxgb/run_experiment_horizontal_tree.sh [hunk 21]
```diff
@@ -1,12 +1,17 @@
 #!/usr/bin/env bash
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo bagging --split_method exponential --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo bagging --split_method exponential --lr_mode scaled --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo bagging --split_method uniform --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo cyclic --split_method exponential --lr_mode uniform --data_split_mode horizontal
-python3 xgb_fl_job_horizontal.py --site_num 5 --training_algo cyclic --split_method uniform --lr_mode uniform --data_split_mode horizontal
+# Tree-based federated XGBoost experiments (bagging and cyclic modes)
+# Now using Recipe API via job_tree.py
```
Comment (chesterxgchen) [inline examples/advanced/xgboost/fedxgb/run_experiment_horizontal_tree.sh]: why do we have a file with just one command

**Review by chesterxgchen — COMMENTED** [review]
Feel like we should avoid create a lot of help files in the example, this make it harder to understand the examples. Try to eliminate––

File: nvflare/app_opt/xgboost/recipes/vertical.py (hunk 18 above)
Comment (YuanTingHsieh) [reply nvflare/app_opt/xgboost/recipes/vertical.py:57]: If differ by just a Boolean I don't think should be 2 classes.
But if more yeah it becomes subjective.

If so can you rename them to be consistent 

XGBHorizontalHistogramRecipe
And
XGBVerticalHistogramRecipe

Or 

XGBHorizontalRecipe
XGBVerticalRecipe
